p2p/discovery/mdns.go

```diff
@@ -37,186 +36,183 @@ type Notifee interface {
 }
 
 type mdnsService struct {
-	server  *mdns.Server
-	service *mdns.MDNSService
-	host    host.Host
-	tag     string
+	host        host.Host
+	serviceName string
+
+	// This ctx is passed to the resolver.
+	// It is closed when Close() is called.
+	ctx       context.Context
+	ctxCancel context.CancelFunc
 
-	lk       sync.Mutex
+	resolverRunning chan struct{}
+	server          *zeroconf.Server
+
+	mutex    sync.Mutex
 	notifees []Notifee
-	interval time.Duration
 }
 
-func getDialableListenAddrs(ph host.Host) ([]*net.TCPAddr, error) {
-	var out []*net.TCPAddr
-	addrs, err := ph.Network().InterfaceListenAddresses()
-	if err != nil {
-		return nil, err
+func NewMdnsService(host host.Host, serviceName string) *mdnsService {
+	ctx, cancel := context.WithCancel(context.Background())
+	if serviceName == "" {
+		serviceName = ServiceName
 	}
+	s := &mdnsService{
+		ctx:             ctx,
+		ctxCancel:       cancel,
+		resolverRunning: make(chan struct{}),
+		host:            host,
+		serviceName:     serviceName,
+	}
+	s.startServer()
+	s.startResolver()
+	return s
+}
+
+func (s *mdnsService) Close() error {
+	s.ctxCancel()
+	if s.server != nil {
+		s.server.Shutdown()
+	}
+	<-s.resolverRunning
+	return nil
+}
+
+// We don't really care about the IP addresses, but the spec (and various routers / firewalls) require us
+// to send A and AAAA records.
+func (s *mdnsService) getIPs(addrs []ma.Multiaddr) ([]string, error) {
+	var ip4, ip6 string
 	for _, addr := range addrs {
-		na, err := manet.ToNetAddr(addr)
+		network, hostport, err := manet.DialArgs(addr)
 		if err != nil {
 			continue
 		}
-		tcp, ok := na.(*net.TCPAddr)
-		if ok {
-			out = append(out, tcp)
+		host, _, err := net.SplitHostPort(hostport)
+		if err != nil {
+			continue
 		}
+		if ip4 == "" && (network == "udp4" || network == "tcp4") {
+			ip4 = host
+		} else if ip6 == "" && (network == "udp6" || network == "tcp6") {
+			ip6 = host
+		}
+	}
+	ips := make([]string, 0, 2)
+	if ip4 != "" {
+		ips = append(ips, ip4)
 	}
-	if len(out) == 0 {
-		return nil, errors.New("failed to find good external addr from peerhost")
+	if ip6 != "" {
+		ips = append(ips, ip6)
 	}
-	return out, nil
+	if len(ips) == 0 {
+		return nil, errors.New("didn't find any IP addresses")
+	}
+	return ips, nil
 }
 
-func NewMdnsService(ctx context.Context, peerhost host.Host, interval time.Duration, serviceTag string) (Service, error) {
-
-	var ipaddrs []net.IP
-	port := 4001
+func (s *mdnsService) mdnsInstance() string {
+	return string(s.host.ID())
+}
 
-	addrs, err := getDialableListenAddrs(peerhost)
+func (s *mdnsService) startServer() error {
+	interfaceAddrs, err := s.host.Network().InterfaceListenAddresses()
```

Comment:
Can we file an issue to fix this? We should be listening to "address change" events, or polling, not advertising a static set of addresses indefinitely.

In terms of MVP, this is no worse than the current system, but we need to fix it.

Reply:
https://github.com/libp2p/go-libp2p/issues/1163